src/System.Private.CoreLib/shared/System/Buffers/MemoryManager.cs

```diff
@@ -11,6 +12,19 @@ namespace System.Buffers
     /// </summary>
     public abstract class MemoryManager<T> : IMemoryOwner<T>, IPinnable
     {
+#pragma warning disable CS0414
+        // It's imperative that this be the first field in the MemoryManager<T> instance
+        // and that it have a negative value. Because it's the first field in the object,
+        // it occupies the same position as String.Length and SzArray.Length; but because
+        // this value is negative and those values are not, we can easily determine whether
+        // any given Memory<T> is backed by a variable-length object (string, array) or
+        // a fixed-length object (MemoryManager).
```

Comment:
Couldn't the runtime decide to rearrange the fields in this class?

Reply:
If we make it a single IntPtr field per your suggestion this should address the rearrangement concern within the base class. The runtime cannot rearrange fields between a base class and it's derived types, so we're good there.

Reply:
Still, this padding is penalizing `MemoryManager` path to make the other path a bit faster. And it is pretty hacky change. I do not think this is a good tradeoff.

Reply:
I had believed the most common usage of `Memory<T>` was to be backed by `String` or `SzArray`. (In fact, I had believed this to be overwhelmingly so.) If this assumption was invalid, then yeah, we shouldn't take the bulk of this PR.

There is still one thing we can take advantage of if we wish: in 32-bit procs, `String` and `SzArray` have the exact same layout. This means we could condition the `if (typeof(T) == typeof(char))` check on 64-bit only, which should give a perf boost to 32-bit procs.

Reply:
> I had believed the most common usage of Memory<T> was to be backed by String or SzArray

I think that's true for workloads like those in ASP.NET.  But I expect that won't be true for certain apps or domains, in particular those that interop with native memory as the lingua franca. The improvements cited for the strings/arrays cases appears to generally be < 10%, whereas the regression for the MemoryManager case looks to be closer to ~40%.

Reply:
> The improvements cited for the strings/arrays cases appears to generally be < 10%, whereas the regression for the MemoryManager case looks to be closer to ~40%.

There are three very specific improvements this PR provides when compared to baseline: (a) improving the performance for String/SzArray, (b) reducing the codegen size by around ~20%, and (c) reducing stack spillage in the calling method due to needing to have temporary Span instances on the stack.

The trade-off for gaining (b) and (c) was to introduce the indirection in the MemoryManager code path. (And remember: even with this regression we're still within 5% of 2.1 RTM since we already made substantial improvements in this code path between 2.1 and 3.0 preview earlier.)

It _is_ possible to make this PR so that we give up (b) and (c). This will improve perf for all three cases: String / SzArray / and MemoryManager. The trade-off is that we reduce the codegen size by only an insignificant amount (~3 - 5%), and we'd still have the dummy `-1` field in MemoryManager.

If this still isn't an appropriate trade-off, that's fine. We can just close the PR - no harm. :)

Reply:
I suppose I shouldn't have said "only an insignificant amount (~3 - 5%)" above. Since this property getter is marked with aggressive inlining it's not unreasonable to think it'd be expanded multiple times in the calling method, so even small improvements would add up.

Reply:
Thanks, @grabyourpitchforks. If I'm understanding you correctly, you're saying there's a version of this change that improves the throughout for strings and arrays equivalent to this one, improves rather than regresses the throughput for MemoryManager, and still shrinks code size some albeit not as much as in this change? That sounds better to me, personally.